Fix UpgradeOnDamage to apply if initial DamageState is valid

reaperrr opened the pull request:
As pointed out by Nolt on IRC, on bleed upgrades don't work even if the initial `DamageState` is valid, until the actor's damage state changes.
However, modders might want to use the trait to apply upgrades while the actor is undamaged, and there might be maps with actors starting at valid damage states, too.

This PR makes the trait perform an additional check for a valid `DamageState` at creation.

reaperrr committed:
Fix granting upgrades on initial damage state

Previously the upgrade(s) would only be granted when the damage state changed, regardless of whether the initial DamageState was already valid.
This prevented the trait from working on Undamaged actors that had just been created, for example.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamage.cs:
```diff
@@ -35,18 +35,37 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 
+		void INotifyCreated.Created(Actor self)
+		{
+			health = self.TraitOrDefault<Health>();
```
Can we put this in the constructor and add `Requires<HealthInfo>`?  It makes zero sense for an actor without Health to implement this trait.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamage.cs:
```diff
@@ -35,18 +35,37 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
+			health = self.TraitOrDefault<Health>();
+
+			if (health == null)
+				throw new YamlException("Actor type '" + self.Info.Name + "': UpgradeOnDamage requires a Health trait!");
```
You can and should then drop this.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamage.cs:
```diff
@@ -35,18 +35,37 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
+			if (health == null)
+				throw new YamlException("Actor type '" + self.Info.Name + "': UpgradeOnDamage requires a Health trait!");
+
+			if (!granted && info.ValidDamageStates.HasFlag(health.DamageState))
```
This bit probably should stay inside `Created`.

reaperrr committed:
Remove bogus 'OrDefault' from UpgradeOnDamage

UpgradeManager is required anyway, so this was redundant.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamage.cs:
```diff
@@ -35,23 +35,35 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		{
-			if (um == null)
-				return;
+			if (!granted && info.ValidDamageStates.HasFlag(health.DamageState))
```
It would be nice to avoid duplicating code here.  Can we get away with calling `DamageStateChanged` directly from here? (note: you'll need to pull out a common private method that both this and `INotifyDamageStateChanged.DamageStateChanged` call)

reaperrr replied:
To be honest, I don't follow, at least not completely. Do you just mean to put the granting stuff into its own method and then just call that method here and in `INotifyDamageStateChanged.DamageStateChanged`, or should `Created` trigger a fake `DamageStateChanged` (pretending it changed)?

GraionDilach replied:
Yeah, I guess he means the former one.

reaperrr committed:
Rename UpgradeOnDamage to UpgradeOnDamageState

The trait works for 'Undamaged' as well and no longer requires a damage state change to trigger either, so this trait name is more accurate.

reaperrr commented:
Updated.
Additionally, renamed trait to `UpgradeOnDamage__State__`, since it supports `Undamaged` and no longer requires the damage state to change to be triggered.

WolfGaming commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamageState.cs:
```diff
@@ -32,41 +32,55 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
-			if (!granted && info.ValidDamageStates.HasFlag(e.DamageState) && !info.ValidDamageStates.HasFlag(e.PreviousDamageState))
+		void GrantUpgradeOnValidDamageState(Actor self, DamageState state)
+		{
+			if (info.ValidDamageStates.HasFlag(state))
```
Isn't the convention to do a not in front and return instead of putting it inside the if here.

WolfGaming commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamageState.cs:
```diff
@@ -32,41 +32,55 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
+			if (!granted && !info.ValidDamageStates.HasFlag(e.PreviousDamageState))
+			{
+				GrantUpgradeOnValidDamageState(self, health.DamageState);
+			}
```
Unneeded curly brackets.

reaperrr replied:
I was a bit torn between convention and readability here (due to the else if below).

WolfGaming replied:
Well I'll leave the final say to @pchote but I have never had these brackets even with the else if,
except when there is an if inside that if.

Phrohdoh replied:
I don't see how these brackets improve readability.

reaperrr commented:
Updated.
I think modders working with releases should get the fix and new name "out of the box", so I'm promoting this to release milestone to make sure it gets merged in time.

WolfGaming commented:
I really feel like those if statements could be made simpler but I don't have any ideas atm.
0,5 :+1: from me until I'm done testing this tonight.

obrakmann commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamageState.cs:
```diff
@@ -32,41 +32,53 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		bool granted;
 
-		public UpgradeOnDamage(Actor self, UpgradeOnDamageInfo info)
+		public UpgradeOnDamageState(Actor self, UpgradeOnDamageStateInfo info)
```
Do we want an upgrade rule for this? This trait has never been released, but it seems that there are already people using it.

reaperrr replied:
I guess it can't hurt.

obrakmann commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamageState.cs:
```diff
@@ -32,41 +32,53 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
-		void INotifyDamageStateChanged.DamageStateChanged(Actor self, AttackInfo e)
+		void INotifyCreated.Created(Actor self)
+		{
+			if (!granted)
```
Isn't this check redundant? I don't see a way how `granted` could become true before `Created` runs.

obrakmann commented:
Looks otherwise fine to me, :+1:

reaperrr committed:
Upgrade rule for UpgradeOnDamage to UpgradeOnDamageState

reaperrr commented:
Updated. Ignore the messed-up commit order, that's just github acting funny again.

Mailaender commented:
:+1:

obrakmann commented:
The test-case can be removed now.

reaperrr commented:
Removed test-case.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ca70eb5043fb6a61ea094546e50dbaa4eab3608f)
